src/cephadm/cephadmlib/runscripts.py

```diff
@@ -0,0 +1,235 @@
+    run_file_path = data_dir / 'unit.run'
+    meta_file_path = data_dir / 'unit.meta'
+    post_stop_file_path = data_dir / 'unit.poststop'
+    stop_file_path = data_dir / 'unit.stop'
+    image_file_path = data_dir / 'unit.image'
+    # use an ExitStack to make writing the files an all-or-nothing affair. If
+    # any file fails to write then the write_new'd file will not get renamed
+    # into place
+    with contextlib.ExitStack() as estack:
+        # write out the main file to run (start) a service
+        runf = estack.enter_context(write_new(run_file_path))
+        runf.write('set -e\n')
+        for command in pre_start_commands or []:
+            _write_command(ctx, runf, command)
+        init_containers = init_containers or []
+        if init_containers:
+            _write_init_container_cmds_clean(ctx, runf, init_containers[0])
+        for idx, ic in enumerate(init_containers):
+            _write_init_container_cmds(ctx, runf, idx, ic)
+        _write_container_cmd_to_bash(ctx, runf, container, ident.daemon_name)
+
+        # some metadata about the deploy
+        metaf = estack.enter_context(write_new(meta_file_path))
+        meta: Dict[str, Any] = fetch_meta(ctx)
+        meta.update(
+            {
+                'memory_request': int(ctx.memory_request)
+                if ctx.memory_request
+                else None,
+                'memory_limit': int(ctx.memory_limit)
+                if ctx.memory_limit
+                else None,
+            }
+        )
+        if not meta.get('ports'):
+            if endpoints:
+                meta['ports'] = [e.port for e in endpoints]
+            else:
+                meta['ports'] = []
+        metaf.write(json.dumps(meta, indent=4) + '\n')
+
+        # post-stop command(s)
+        pstopf = estack.enter_context(write_new(post_stop_file_path))
+        # this is a fallback to eventually stop any underlying container that
+        # was not stopped properly by unit.stop, this could happen in very slow
+        # setups as described in the issue
+        # https://tracker.ceph.com/issues/58242.
+        _write_stop_actions(ctx, cast(TextIO, pstopf), container, timeout)
+        for command in post_stop_commands or []:
+            _write_command(ctx, pstopf, command)
+
+        # stop command(s)
+        stopf = estack.enter_context(write_new(stop_file_path))
+        _write_stop_actions(ctx, cast(TextIO, stopf), container, timeout)
+
+        if container:
+            imgf = estack.enter_context(write_new(image_file_path))
+            imgf.write(container.image + '\n')
+
+
+def _write_container_cmd_to_bash(
+    ctx: CephadmContext,
+    file_obj: IO[str],
+    container: 'CephContainer',
+    comment: Optional[str] = None,
+    background: Optional[bool] = False,
+) -> None:
+    if comment:
+        # Sometimes adding a comment, especially if there are multiple containers in one
+        # unit file, makes it easier to read and grok.
```

Comment:
Was going to ask what `grok` means, but it looks like it was an existing comment being moved.

Reply:
That's true - I didn't write the original comment. I don't typically use it actively but I do know what it means though :-)
It's a "Martian" word that roughly means to understand, invented by writer Robert Heinlein that caught on outside the book he created it for. https://www.wordnik.com/words/grok

Reply:
... and yes, I've read the book.